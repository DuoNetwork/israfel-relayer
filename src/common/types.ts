import { SignedOrder } from '0x.js';

export interface ILiveOrder {
	account: string;
	pair: string;
	orderHash: string;
	price: number;
	amount: number;
	balance: number;
	fill: number;
	side: string;
	createdAt?: number;
	updatedAt?: number;
	initialSequence: number;
	currentSequence: number;
}

export interface IUserOrder extends ILiveOrder {
	type: string;
	status: string;
	updatedBy: string;
}

export interface IRawOrder {
	orderHash: string;
	signedOrder: IStringSignedOrder | SignedOrder;
	createdAt?: number;
	updatedAt?: number;
}

export interface IStringSignedOrder {
	senderAddress: string;
	makerAddress: string;
	takerAddress: string;
	makerFee: string;
	takerFee: string;
	makerAssetAmount: string;
	takerAssetAmount: string;
	makerAssetData: string;
	takerAssetData: string;
	salt: string;
	exchangeAddress: string;
	feeRecipientAddress: string;
	expirationTimeSeconds: string;
	signature: string;
}

export interface IOrderPersistRequest {
	method: string;
	pair: string;
	orderHash: string;
	balance: number;
	signedOrder?: IStringSignedOrder;
}

export interface IOrderQueueItem {
	method: string;
	liveOrder: ILiveOrder;
	signedOrder?: IStringSignedOrder;
}

export interface IOrderBookSnapshotWs extends IOrderBookSnapshot {
	type: string;
	sequence: number;
	channel: string;
}

export interface IOrderBookSnapshot {
	sequence: number;
	bids: IOrderBookUpdateWS[];
	asks: IOrderBookUpdateWS[];
}

export interface IUpdatePayloadWs {
	order: IStringSignedOrder;
	metaData: {
		remainingTakerAssetAmount: string;
	};
}

export interface IOrderBookUpdateWS {
	price: number;
	amount: number;
}

export interface IOrderBookUpdate {
	sequence: number;
	pair: string;
	price: number;
	amount: number;
}

export interface IWsRequest {
	method: string;
	channel: string;
	pair: string;
}

export interface IWsResponse {
	status: string;
	method: string;
	channel: string;
	pair: string;
}

export interface IWsOrderRequest extends IWsRequest {
	orderHash: string;
}

export interface IWsAddOrderRequest extends IWsOrderRequest {
	order: IStringSignedOrder | SignedOrder;
}

export interface IWsOrderResponse extends IWsResponse {
	orderHash: string;
}

export interface IWsUserOrderResponse extends IWsOrderResponse {
	userOrder: IUserOrder;
}

export interface ISubscribeOrderBookRequest {
	method: string;
	channel: string;
}

// export interface IUpdateResponseWs {
// 	type: WsChannelResposnseTypes;
// 	lastTimestamp: number;
// 	currentTimestamp: number;
// 	channel: {
// 		name: WsChannelName;
// 		pair: string;
// 	};
// 	bids: IOrderBookUpdateWS[];
// 	asks: IOrderBookUpdateWS[];
// }

// export interface IOrderResponseWs {
// 	channel: {
// 		name: WsChannelName;
// 		pair: string;
// 	};
// 	status: string;
// 	failedReason: string;
// }

export interface IOption {
	live: boolean;
	token: string;
	maker: number;
	spender: number;
	amount: number;
	debug: boolean;
	server: boolean;
}

export interface IStatus {
	hostname: string;
	updatedAt: number;
	pair: string;
	tool: string;
	count?: number;
}

export interface IService {
	service: string;
	hostname: string;
	url: string;
}
