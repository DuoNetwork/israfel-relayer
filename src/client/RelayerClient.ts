import WebSocket from 'isomorphic-ws';
import * as CST from '../common/constants';
import {
	IAcceptedPrice,
	IOrderBookSnapshot,
	IOrderBookSnapshotUpdate,
	IPrice,
	IStatus,
	IToken,
	IUserOrder,
	IWsAddOrderRequest,
	IWsInfoResponse,
	IWsOrderBookResponse,
	IWsOrderBookUpdateResponse,
	IWsOrderHistoryRequest,
	IWsOrderHistoryResponse,
	IWsOrderResponse,
	IWsRequest,
	IWsResponse,
	IWsTerminateOrderRequest,
	IWsUserOrderResponse
} from '../common/types';
import orderBookUtil from '../utils/orderBookUtil';
import orderUtil from '../utils/orderUtil';
import Web3Util from '../utils/Web3Util';

export default class RelayerClient {
	public web3Util: Web3Util;
	private env: string;
	public ws: WebSocket | null = null;
	private handleConnected: () => any = () => ({});
	private handleReconnect: () => any = () => ({});
	private handleInfoUpdate: (
		tokens: IToken[],
		status: IStatus[],
		acceptedPrices: { [custodian: string]: IAcceptedPrice[] },
		exchangePrices: { [source: string]: IPrice[] }
	) => any = () => ({});
	private handleOrderUpdate: (userOrder: IUserOrder) => any = () => ({});
	private handleOrderHistoryUpdate: (userOrders: IUserOrder[]) => any = () => ({});
	private handleOrderError: (
		method: string,
		orderHash: string,
		error: string
	) => any = () => ({});
	private handleOrderBookUpdate: (orderBookSnapshot: IOrderBookSnapshot) => any = () => ({});
	private handleOrderBookError: (method: string, pair: string, error: string) => any = () => ({});
	public reconnectionNumber: number = 0;
	public orderBookSnapshots: { [pair: string]: IOrderBookSnapshot } = {};
	public pendingOrderBookUpdates: { [pair: string]: IOrderBookSnapshotUpdate[] } = {};
	public orderBookSnapshotAvailable: { [pair: string]: boolean } = {};

	constructor(web3Util: Web3Util, env: string) {
		this.web3Util = web3Util;
		this.env = env;
	}

	private reconnect() {
		if (this.ws) {
			this.ws.removeAllListeners();
			this.ws.terminate();
		}
		this.ws = null;
		this.handleReconnect();
		if (this.reconnectionNumber < 5) {
			this.reconnectionNumber++;
			setTimeout(() => this.connectToRelayer(), this.reconnectionNumber * 10000);
		}
	}

	public connectToRelayer() {
		this.ws = new WebSocket(`wss://relayer.${this.env}.israfel.info:8080`);
		this.ws.onopen = () => {
			this.reconnectionNumber = 0;
			this.handleConnected();
		};
		this.ws.onmessage = (m: any) => this.handleMessage(m.data.toString());
		this.ws.onerror = error => {
			console.log('ws error');
			console.log(JSON.stringify(error));
			this.reconnect();
		};
		this.ws.onclose = () => {
			console.log('ws close');
			this.reconnect();
		};
	}

	public handleOrderResponse(response: IWsResponse) {
		if (response.status !== CST.WS_OK)
			this.handleOrderError(
				response.method,
				(response as IWsOrderResponse).orderHash,
				response.status
			);
		else if (response.method === CST.WS_HISTORY)
			this.handleOrderHistoryUpdate((response as IWsOrderHistoryResponse).orderHistory);
		else this.handleOrderUpdate((response as IWsUserOrderResponse).userOrder);
	}

	public handleOrderBookResponse(orderBookResponse: IWsResponse) {
		if (orderBookResponse.status !== CST.WS_OK)
			this.handleOrderBookError(
				orderBookResponse.method,
				orderBookResponse.pair,
				orderBookResponse.status
			);
		else if (orderBookResponse.method === CST.DB_SNAPSHOT) {
			const { pair, orderBookSnapshot } = orderBookResponse as IWsOrderBookResponse;
			this.orderBookSnapshots[pair] = orderBookSnapshot;
			const pendingUpdates = this.pendingOrderBookUpdates[pair];
			let hasGap = false;
			if (pendingUpdates && pendingUpdates.length) {
				pendingUpdates.sort((a, b) => a.version - b.version);
				while (pendingUpdates.length) {
					const update = pendingUpdates.shift();
					if (!update) break;
					if (update.prevVersion < orderBookSnapshot.version) continue;
					if (update.prevVersion > orderBookSnapshot.version) {
						hasGap = true;
						pendingUpdates.unshift(update);
						break;
					}
					orderBookUtil.updateOrderBookSnapshot(orderBookSnapshot, update);
				}
			}

			if (!hasGap) this.orderBookSnapshotAvailable[pair] = true;
			else this.subscribeOrderBook(pair);

			this.handleOrderBookUpdate(orderBookSnapshot);
		} else if (orderBookResponse.method === CST.DB_UPDATE) {
			const { pair, orderBookUpdate } = orderBookResponse as IWsOrderBookUpdateResponse;
			if (!this.orderBookSnapshotAvailable[pair])
				this.pendingOrderBookUpdates[pair].push(orderBookUpdate);
			else {
				const orderBookSnapshot = this.orderBookSnapshots[pair];
				if (orderBookUpdate.prevVersion === orderBookSnapshot.version) {
					orderBookUtil.updateOrderBookSnapshot(orderBookSnapshot, orderBookUpdate);
					this.handleOrderBookUpdate(orderBookSnapshot);
				} else if (orderBookUpdate.prevVersion > orderBookSnapshot.version) {
					this.pendingOrderBookUpdates[pair].push(orderBookUpdate);
					this.subscribeOrderBook(pair);
				}
			}
		}
	}

	public handleMessage(message: string) {
		const res: IWsResponse = JSON.parse(message);
		if (res.method !== CST.WS_UNSUB)
			switch (res.channel) {
				case CST.DB_ORDERS:
					this.handleOrderResponse(res as IWsOrderResponse);
					break;
				case CST.DB_ORDER_BOOKS:
					this.handleOrderBookResponse(res);
					break;
				case CST.WS_INFO:
					const {
						tokens,
						processStatus,
						acceptedPrices,
						exchangePrices
					} = res as IWsInfoResponse;
					this.web3Util.setTokens(tokens);
					this.handleInfoUpdate(tokens, processStatus, acceptedPrices, exchangePrices);
					break;
				default:
					break;
			}
	}

	public subscribeOrderBook(pair: string) {
		if (!this.ws) return;
		this.orderBookSnapshotAvailable[pair] = false;
		if (!this.pendingOrderBookUpdates[pair]) this.pendingOrderBookUpdates[pair] = [];
		const msg: IWsRequest = {
			method: CST.WS_SUB,
			channel: CST.DB_ORDER_BOOKS,
			pair: pair
		};
		this.ws.send(JSON.stringify(msg));
	}

	public unsubscribeOrderBook(pair: string) {
		if (!this.ws) return;

		const msg: IWsRequest = {
			method: CST.WS_UNSUB,
			channel: CST.DB_ORDER_BOOKS,
			pair: pair
		};
		this.ws.send(JSON.stringify(msg));
		this.orderBookSnapshotAvailable[pair] = false;
		delete this.orderBookSnapshots[pair];
		this.pendingOrderBookUpdates[pair] = [];
	}

	public subscribeOrderHistory(account: string) {
		if (!this.ws) return;

		if (!Web3Util.isValidAddress(account)) return;

		const msg: IWsOrderHistoryRequest = {
			method: CST.WS_SUB,
			channel: CST.DB_ORDERS,
			pair: '',
			account: account
		};
		this.ws.send(JSON.stringify(msg));
	}

	public unsubscribeOrderHistory(account: string) {
		if (!this.ws) return;

		const msg: IWsOrderHistoryRequest = {
			method: CST.WS_UNSUB,
			channel: CST.DB_ORDERS,
			pair: '',
			account: account
		};
		this.ws.send(JSON.stringify(msg));
	}

	public async addOrder(
		account: string,
		pair: string,
		price: number,
		amount: number,
		isBid: boolean,
		expiry: number
	) {
		if (!this.ws) return '';
		if (!this.web3Util.isValidPair(pair)) throw new Error('invalid pair');
		const [code1, code2] = pair.split('|');
		const token1 = this.web3Util.getTokenByCode(code1);
		if (!token1) throw new Error('invalid pair');
		const address1 = token1.address;
		const address2 = this.web3Util.getTokenAddressFromCode(code2);
		const amountAfterFee = orderUtil.getAmountAfterFee(
			amount,
			price,
			token1.feeSchedules[code2],
			isBid
		);

		if (!amountAfterFee.makerAssetAmount || !amountAfterFee.takerAssetAmount)
			throw new Error('invalid amount');

		const rawOrder = await this.web3Util.createRawOrder(
			pair,
			account,
			isBid ? address2 : address1,
			isBid ? address1 : address2,
			amountAfterFee.makerAssetAmount,
			amountAfterFee.takerAssetAmount,
			Math.ceil(expiry / 1000)
		);
		const msg: IWsAddOrderRequest = {
			method: CST.DB_ADD,
			channel: CST.DB_ORDERS,
			pair: pair,
			orderHash: rawOrder.orderHash,
			order: rawOrder.signedOrder
		};
		this.ws.send(JSON.stringify(msg));
		return rawOrder.orderHash;
	}

	public deleteOrder(pair: string, orderHashes: string[], signature: string) {
		if (!this.ws) return;

		const msg: IWsTerminateOrderRequest = {
			method: CST.DB_TERMINATE,
			channel: CST.DB_ORDERS,
			pair: pair,
			orderHashes: orderHashes,
			signature: signature
		};
		this.ws.send(JSON.stringify(msg));
	}

	public onOrder(
		handleHistory: (userOrders: IUserOrder[]) => any,
		handleUpdate: (userOrder: IUserOrder) => any,
		handleError: (method: string, orderHash: string, error: string) => any
	) {
		this.handleOrderHistoryUpdate = handleHistory;
		this.handleOrderUpdate = handleUpdate;
		this.handleOrderError = handleError;
	}

	public onOrderBook(
		handleUpdate: (orderBookSnapshot: IOrderBookSnapshot) => any,
		handleError: (method: string, pair: string, error: string) => any
	) {
		this.handleOrderBookUpdate = handleUpdate;
		this.handleOrderBookError = handleError;
	}

	public onConnection(handleConnected: () => any, handleReconnect: () => any) {
		this.handleConnected = handleConnected;
		this.handleReconnect = handleReconnect;
	}

	public onInfoUpdate(
		handleInfoUpdate: (
			tokens: IToken[],
			status: IStatus[],
			acceptedPrices: { [custodian: string]: IAcceptedPrice[] },
			exchangePrices: { [source: string]: IPrice[] }
		) => any
	) {
		this.handleInfoUpdate = handleInfoUpdate;
	}
}
