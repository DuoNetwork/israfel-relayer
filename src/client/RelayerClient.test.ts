import * as CST from '../common/constants';
import { IWsOrderBookResponse, IWsOrderBookUpdateResponse } from '../common/types';
import orderBookUtil from '../utils/orderBookUtil';
import orderUtil from '../utils/orderUtil';
// import Web3Util from '../utils/Web3Util';
import RelayerClient from './RelayerClient';

const web3Util: any = {};
const relayerClient = new RelayerClient(web3Util as any, CST.DB_DEV);

test('handleOrderResponse ok', () => {
	const handleUpdate = jest.fn();
	const handleError = jest.fn();
	const handleHistory = jest.fn();
	relayerClient.onOrder(handleHistory, handleUpdate, handleError);
	relayerClient.handleOrderResponse({
		channel: 'channel',
		method: 'method',
		status: CST.WS_OK,
		orderHash: '0xOrderHash',
		pair: 'pair',
		userOrder: 'userOrder'
	} as any);
	expect(handleUpdate.mock.calls).toMatchSnapshot();
	expect(handleHistory).not.toBeCalled();
	expect(handleError).not.toBeCalled();
});

test('handleOrderResponse history', () => {
	const handleUpdate = jest.fn();
	const handleError = jest.fn();
	const handleHistory = jest.fn();
	relayerClient.onOrder(handleHistory, handleUpdate, handleError);
	relayerClient.handleOrderResponse({
		channel: 'channel',
		method: CST.WS_HISTORY,
		status: CST.WS_OK,
		pair: 'pair',
		orderHistory: 'orderHistory'
	} as any);
	expect(handleUpdate).not.toBeCalled();
	expect(handleHistory.mock.calls).toMatchSnapshot();
	expect(handleError).not.toBeCalled();
});

test('handleOrderResponse not ok', () => {
	const handleUpdate = jest.fn();
	const handleError = jest.fn();
	const handleHistory = jest.fn();
	relayerClient.onOrder(handleHistory, handleUpdate, handleError);
	relayerClient.handleOrderResponse({
		channel: 'channel',
		method: 'method',
		status: 'status',
		orderHash: '0xOrderHash',
		pair: 'pair'
	} as any);
	expect(handleUpdate).not.toBeCalled();
	expect(handleHistory).not.toBeCalled();
	expect(handleError.mock.calls).toMatchSnapshot();
});

test('handleOrderBookResponse update before snapshot', () => {
	const handleUpdate = jest.fn();
	const handleError = jest.fn();
	relayerClient.onOrderBook(handleUpdate, handleError);
	const res: IWsOrderBookUpdateResponse = {
		channel: 'channel',
		status: CST.WS_OK,
		method: CST.DB_UPDATE,
		pair: 'pair',
		orderBookUpdate: {
			pair: 'pair',
			updates: [],
			prevVersion: 122,
			version: 123
		}
	};
	relayerClient.handleOrderBookResponse(res);
	expect(handleUpdate).not.toBeCalled();
	expect(handleError).not.toBeCalled();
	expect(relayerClient.pendingOrderBookUpdates).toMatchSnapshot();
});

test('handleOrderBookResponse snapshot newer than pending updates', () => {
	const handleUpdate = jest.fn();
	const handleError = jest.fn();
	relayerClient.onOrderBook(handleUpdate, handleError);
	orderBookUtil.updateOrderBookSnapshot = jest.fn();
	const res: IWsOrderBookResponse = {
		channel: 'channel',
		status: CST.WS_OK,
		method: CST.DB_SNAPSHOT,
		pair: 'pair',
		orderBookSnapshot: {
			pair: 'pair',
			version: 124,
			bids: [],
			asks: [],
		}
	}
	relayerClient.handleOrderBookResponse(res);
	expect(handleUpdate).toBeCalled();
	expect(handleError).not.toBeCalled();
	expect(orderBookUtil.updateOrderBookSnapshot as jest.Mock).not.toBeCalled();
	expect(relayerClient.pendingOrderBookUpdates['pair']).toEqual([]);
	expect(relayerClient.orderBookSnapshotAvailable['pair']).toBeTruthy();
});

test('handleOrderBookResponse snapshot older than pending updates', () => {
	const handleUpdate = jest.fn();
	const handleError = jest.fn();
	relayerClient.onOrderBook(handleUpdate, handleError);
	orderBookUtil.updateOrderBookSnapshot = jest.fn();
	relayerClient.pendingOrderBookUpdates['pair'] = [ {
		pair: 'pair',
		updates: [],
		prevVersion: 122,
		version: 123
	}];
	const res: IWsOrderBookResponse = {
		channel: 'channel',
		status: CST.WS_OK,
		method: CST.DB_SNAPSHOT,
		pair: 'pair',
		orderBookSnapshot: {
			pair: 'pair',
			version: 122,
			bids: [],
			asks: [],
		}
	}
	relayerClient.handleOrderBookResponse(res);
	expect(handleUpdate).toBeCalled();
	expect(handleError).not.toBeCalled();
	expect((orderBookUtil.updateOrderBookSnapshot as jest.Mock).mock.calls).toMatchSnapshot();
	expect(relayerClient.pendingOrderBookUpdates['pair']).toEqual([]);
	expect(relayerClient.orderBookSnapshotAvailable['pair']).toBeTruthy();
});

test('handleOrderBookResponse update after snapshot', () => {
	const handleUpdate = jest.fn();
	const handleError = jest.fn();
	relayerClient.onOrderBook(handleUpdate, handleError);
	orderBookUtil.updateOrderBookSnapshot = jest.fn();
	const res: IWsOrderBookUpdateResponse = {
		channel: 'channel',
		status: CST.WS_OK,
		method: CST.DB_UPDATE,
		pair: 'pair',
		orderBookUpdate: {
			pair: 'pair',
			updates: [],
			prevVersion: 122,
			version: 123
		}
	};
	relayerClient.handleOrderBookResponse(res);
	expect(handleUpdate).toBeCalled();
	expect(handleError).not.toBeCalled();
	expect((orderBookUtil.updateOrderBookSnapshot as jest.Mock).mock.calls).toMatchSnapshot();
	expect(relayerClient.pendingOrderBookUpdates['pair']).toEqual([]);
	expect(relayerClient.orderBookSnapshotAvailable['pair']).toBeTruthy();
});

test('handleOrderBookResponse not ok', () => {
	const handleUpdate = jest.fn();
	const handleError = jest.fn();
	relayerClient.onOrderBook(handleUpdate, handleError);
	relayerClient.handleOrderBookResponse({
		channel: 'channel',
		status: 'status',
		method: 'any',
		pair: 'pair',
		orderBookUpdate: 'orderBookUpdate'
	} as any);
	expect(handleUpdate).not.toBeCalled();
	expect(handleError.mock.calls).toMatchSnapshot();
});

test('handleMessage unsub', () => {
	relayerClient.handleOrderResponse = jest.fn();
	relayerClient.handleOrderBookResponse = jest.fn();
	const handleInfo = jest.fn();
	relayerClient.onInfoUpdate(handleInfo);
	relayerClient.handleMessage(
		JSON.stringify({
			method: CST.WS_UNSUB
		})
	);
	expect(relayerClient.handleOrderResponse as jest.Mock).not.toBeCalled();
	expect(relayerClient.handleOrderBookResponse as jest.Mock).not.toBeCalled();
	expect(handleInfo).not.toBeCalled();
});

test('handleMessage invalid channel', () => {
	relayerClient.handleOrderResponse = jest.fn();
	relayerClient.handleOrderBookResponse = jest.fn();
	const handleInfo = jest.fn();
	relayerClient.onInfoUpdate(handleInfo);
	relayerClient.handleMessage(
		JSON.stringify({
			channel: 'channel'
		})
	);
	expect(relayerClient.handleOrderResponse as jest.Mock).not.toBeCalled();
	expect(relayerClient.handleOrderBookResponse as jest.Mock).not.toBeCalled();
	expect(handleInfo).not.toBeCalled();
});

test('handleMessage orders', () => {
	relayerClient.handleOrderResponse = jest.fn();
	relayerClient.handleOrderBookResponse = jest.fn();
	const handleInfo = jest.fn();
	relayerClient.onInfoUpdate(handleInfo);
	relayerClient.handleMessage(
		JSON.stringify({
			channel: CST.DB_ORDERS
		})
	);
	expect((relayerClient.handleOrderResponse as jest.Mock).mock.calls).toMatchSnapshot();
	expect(relayerClient.handleOrderBookResponse as jest.Mock).not.toBeCalled();
	expect(handleInfo).not.toBeCalled();
});

test('handleMessage orderBooks', () => {
	relayerClient.handleOrderResponse = jest.fn();
	relayerClient.handleOrderBookResponse = jest.fn();
	const handleInfo = jest.fn();
	relayerClient.onInfoUpdate(handleInfo);
	relayerClient.handleMessage(
		JSON.stringify({
			channel: CST.DB_ORDER_BOOKS
		})
	);
	expect(relayerClient.handleOrderResponse as jest.Mock).not.toBeCalled();
	expect((relayerClient.handleOrderBookResponse as jest.Mock).mock.calls).toMatchSnapshot();
	expect(handleInfo).not.toBeCalled();
});

test('handleMessage info', () => {
	web3Util.setTokens = jest.fn();
	relayerClient.handleOrderResponse = jest.fn();
	relayerClient.handleOrderBookResponse = jest.fn();
	const handleInfo = jest.fn();
	relayerClient.onInfoUpdate(handleInfo);
	relayerClient.handleMessage(
		JSON.stringify({
			channel: CST.WS_INFO,
			tokens: 'tokens',
			processStatus: 'status',
			acceptedPrices: 'acceptedPrices',
			exchangePrices: 'exchangePrices'
		})
	);
	expect(relayerClient.handleOrderResponse as jest.Mock).not.toBeCalled();
	expect(relayerClient.handleOrderBookResponse as jest.Mock).not.toBeCalled();
	expect((web3Util.setTokens as jest.Mock).mock.calls).toMatchSnapshot();
	expect(handleInfo.mock.calls).toMatchSnapshot();
});

test('subscribeOrderBook', () => {
	const send = jest.fn();
	relayerClient.ws = { send } as any;
	relayerClient.subscribeOrderBook('pair');
	expect(send.mock.calls).toMatchSnapshot();
});

test('unsubscribeOrderBook', () => {
	const send = jest.fn();
	relayerClient.ws = { send } as any;
	relayerClient.unsubscribeOrderBook('pair');
	expect(send.mock.calls).toMatchSnapshot();
});

test('subscribeOrderHistory invalid address', () => {
	const send = jest.fn();
	relayerClient.ws = { send } as any;
	relayerClient.subscribeOrderHistory('account');
	expect(send).not.toBeCalled();
});

test('subscribeOrderHistory', () => {
	const send = jest.fn();
	relayerClient.ws = { send } as any;
	relayerClient.subscribeOrderHistory('0x48bacb9266a570d521063ef5dd96e61686dbe788');
	expect(send.mock.calls).toMatchSnapshot();
});

test('unsubscribeOrderHistory', () => {
	const send = jest.fn();
	relayerClient.ws = { send } as any;
	relayerClient.unsubscribeOrderHistory('account');
	expect(send.mock.calls).toMatchSnapshot();
});

test('addOrder bid', async done => {
	const send = jest.fn();
	relayerClient.ws = { send } as any;
	web3Util.getTokenAddressFromCode = jest.fn((code: string) => code + 'address');
	orderUtil.getAmountAfterFee = jest.fn(() => ({
		takerAssetAmount: 123,
		makerAssetAmount: 456
	}));
	web3Util.isValidPair = jest.fn(() => true);
	web3Util.getTokenByCode = jest.fn(() => ({
		address: 'code1address',
		code: 'code1',
		precisions: {
			code2: 1
		},
		feeSchedules: {
			code2: {}
		}
	}));
	web3Util.createRawOrder = jest.fn(() => ({
		orderHash: 'orderHash',
		signedOrder: 'signedOrder'
	}));
	await relayerClient.addOrder('account', 'code1|code2', 123, 456, true, 1234567890);
	expect(send.mock.calls).toMatchSnapshot();
	expect((orderUtil.getAmountAfterFee as jest.Mock).mock.calls).toMatchSnapshot();
	expect((web3Util.createRawOrder as jest.Mock).mock.calls).toMatchSnapshot();
	done();
});

test('addOrder ask', async done => {
	const send = jest.fn();
	relayerClient.ws = { send } as any;
	web3Util.getTokenAddressFromCode = jest.fn((code: string) => code + 'address');
	orderUtil.getAmountAfterFee = jest.fn(() => ({
		takerAssetAmount: 123,
		makerAssetAmount: 456
	}));
	web3Util.isValidPair = jest.fn(() => true);
	web3Util.getTokenByCode = jest.fn(() => ({
		address: 'code1address',
		code: 'code1',
		precisions: {
			code2: 1
		},
		feeSchedules: {
			code2: {}
		}
	}));
	web3Util.createRawOrder = jest.fn(() => ({
		orderHash: 'orderHash',
		signedOrder: 'signedOrder'
	}));
	await relayerClient.addOrder('account', 'code1|code2', 123, 456, false, 1234567890);
	expect(send.mock.calls).toMatchSnapshot();
	expect((orderUtil.getAmountAfterFee as jest.Mock).mock.calls).toMatchSnapshot();
	expect((web3Util.createRawOrder as jest.Mock).mock.calls).toMatchSnapshot();
	done();
});

test('deleteOrder', () => {
	const send = jest.fn();
	relayerClient.ws = { send } as any;
	relayerClient.deleteOrder('pair', 'orderHash', 'signature');
	expect(send.mock.calls).toMatchSnapshot();
});
