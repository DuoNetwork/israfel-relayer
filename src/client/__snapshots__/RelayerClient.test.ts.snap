// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addOrder ask 1`] = `
Array [
  Array [
    "{\\"method\\":\\"add\\",\\"channel\\":\\"orders\\",\\"pair\\":\\"code1|code2\\",\\"orderHash\\":\\"orderHash\\",\\"order\\":\\"signedOrder\\"}",
  ],
]
`;

exports[`addOrder ask 2`] = `
Array [
  Array [
    456,
    123,
    Object {},
    false,
  ],
]
`;

exports[`addOrder ask 3`] = `
Array [
  Array [
    "code1|code2",
    "account",
    "code1address",
    "code2address",
    456,
    123,
    1234568,
  ],
]
`;

exports[`addOrder bid 1`] = `
Array [
  Array [
    "{\\"method\\":\\"add\\",\\"channel\\":\\"orders\\",\\"pair\\":\\"code1|code2\\",\\"orderHash\\":\\"orderHash\\",\\"order\\":\\"signedOrder\\"}",
  ],
]
`;

exports[`addOrder bid 2`] = `
Array [
  Array [
    456,
    123,
    Object {},
    true,
  ],
]
`;

exports[`addOrder bid 3`] = `
Array [
  Array [
    "code1|code2",
    "account",
    "code2address",
    "code1address",
    456,
    123,
    1234568,
  ],
]
`;

exports[`addOrder invalid amount 1`] = `[Error: invalid amount]`;

exports[`addOrder invalid pair 1`] = `[Error: invalid pair]`;

exports[`addOrder missing token 1`] = `[Error: invalid pair]`;

exports[`deleteOrder 1`] = `
Array [
  Array [
    "{\\"method\\":\\"terminate\\",\\"channel\\":\\"orders\\",\\"pair\\":\\"pair\\",\\"orderHashes\\":[\\"orderHash\\"],\\"signature\\":\\"signature\\"}",
  ],
]
`;

exports[`handleMessage info 1`] = `
Array [
  Array [
    "tokens",
  ],
]
`;

exports[`handleMessage info 2`] = `
Array [
  Array [
    "tokens",
    "status",
    "acceptedPrices",
    "exchangePrices",
  ],
]
`;

exports[`handleMessage orderBooks 1`] = `
Array [
  Array [
    Object {
      "channel": "orderBooks",
    },
  ],
]
`;

exports[`handleMessage orders 1`] = `
Array [
  Array [
    Object {
      "channel": "orders",
    },
  ],
]
`;

exports[`handleOrderBookResponse not ok 1`] = `
Array [
  Array [
    "any",
    "pair",
    "status",
  ],
]
`;

exports[`handleOrderBookResponse snapshot newer than pending updates 1`] = `
Array [
  Array [
    Object {
      "asks": Array [],
      "bids": Array [],
      "pair": "pair",
      "version": 124,
    },
  ],
]
`;

exports[`handleOrderBookResponse snapshot no pending updates 1`] = `
Array [
  Array [
    Object {
      "asks": Array [],
      "bids": Array [],
      "pair": "pair",
      "version": 122,
    },
  ],
]
`;

exports[`handleOrderBookResponse snapshot older than pending updates 1`] = `
Array [
  Array [
    Object {
      "asks": Array [],
      "bids": Array [],
      "pair": "pair",
      "version": 123,
    },
  ],
]
`;

exports[`handleOrderBookResponse snapshot older than pending updates but has gap 1`] = `
Array [
  Array [
    Object {
      "asks": Array [],
      "bids": Array [],
      "pair": "pair",
      "version": 123,
    },
  ],
]
`;

exports[`handleOrderBookResponse snapshot older than pending updates but has gap 2`] = `
Array [
  Object {
    "pair": "pair",
    "prevVersion": 124,
    "updates": Array [],
    "version": 125,
  },
]
`;

exports[`handleOrderBookResponse update after snapshot 1`] = `
Array [
  Array [
    Object {
      "asks": Array [],
      "bids": Array [],
      "pair": "pair",
      "version": 123,
    },
  ],
]
`;

exports[`handleOrderBookResponse update after snapshot has gap 1`] = `
Array [
  Object {
    "pair": "pair",
    "prevVersion": 123,
    "updates": Array [],
    "version": 124,
  },
]
`;

exports[`handleOrderBookResponse update before snapshot 1`] = `
Object {
  "pair": Array [
    Object {
      "pair": "pair",
      "prevVersion": 122,
      "updates": Array [],
      "version": 123,
    },
  ],
}
`;

exports[`handleOrderResponse history 1`] = `
Array [
  Array [
    "orderHistory",
  ],
]
`;

exports[`handleOrderResponse not ok 1`] = `
Array [
  Array [
    "method",
    "0xOrderHash",
    "status",
  ],
]
`;

exports[`handleOrderResponse ok 1`] = `
Array [
  Array [
    "userOrder",
  ],
]
`;

exports[`subscribeOrderBook existing pair 1`] = `
Array [
  Array [
    "{\\"method\\":\\"subscribe\\",\\"channel\\":\\"orderBooks\\",\\"pair\\":\\"pair\\"}",
  ],
]
`;

exports[`subscribeOrderBook new pair 1`] = `
Array [
  Array [
    "{\\"method\\":\\"subscribe\\",\\"channel\\":\\"orderBooks\\",\\"pair\\":\\"pair\\"}",
  ],
]
`;

exports[`subscribeOrderHistory 1`] = `
Array [
  Array [
    "{\\"method\\":\\"subscribe\\",\\"channel\\":\\"orders\\",\\"pair\\":\\"\\",\\"account\\":\\"0x48bacb9266a570d521063ef5dd96e61686dbe788\\"}",
  ],
]
`;

exports[`unsubscribeOrderBook 1`] = `
Array [
  Array [
    "{\\"method\\":\\"unsubscribe\\",\\"channel\\":\\"orderBooks\\",\\"pair\\":\\"pair\\"}",
  ],
]
`;

exports[`unsubscribeOrderHistory 1`] = `
Array [
  Array [
    "{\\"method\\":\\"unsubscribe\\",\\"channel\\":\\"orders\\",\\"pair\\":\\"\\",\\"account\\":\\"account\\"}",
  ],
]
`;
